[typescript] Add tests for typings + fixes

sebald opened the pull request:
This patch will add testing for TS typings (finally) and fixes some wrong typings (details below). 

Closes #7696

- [x] Ship the typings with the build
- [x] Run the tests `yarn run test:typings [-- --watch]`
- [x] Add typings for `GridList`
- [x] Add typings for transitions `Collapse`, `Fade` and `Slide`
- [x] Fix `TestFieldProps` (typo...)
- [x] Fix `SnackbarProps` (`action` can be an array, `message` can be a string)
- [x] Fix `TransitionProps` (`duration` can be a string)
- [x] Fix `FormGroupProps` (`row` is optional)
- [x] Fix `SlideProps` (`direction` + `offset` where missing)
- [x] Use `SlideProps` in `DrawerProps`

---

- [x] PR has tests / docs demo, and is linted.
- [x] Commit and PR titles begin with [ComponentName], and are in imperative form: "[Component] Fix leaky abstraction".
- [x] Description explains the issue / use-case resolved, and auto-closes the related issue(s) (http://tr.im/vFqem).

sebald commented:
@oliviertassinari I will move the `index.d.ts` to the `src` folder. But where should I put the tests? :)

oliviertassinari commented:
@sebald You can put the tests under the `/test/typescript`. Do you have another suggestion?

sebald commented:
@oliviertassinari No, just wasn't sure if you like to have the typing tests under `test`:)

oliviertassinari commented:
@sebald Should we run the typescript tests in the CI?

sebald commented:
@oliviertassinari Would be nice. But it shouldn't hold back a new release, I guess.

shousper commented:
Hi @sebald,

If you wouldn't mind including a similar fix to the one found in #7697 in this PR before it's merged, that'd be greatly appreciated! :)

At the moment it appears the `/typings` folder isn't actually included in the package published to npm, nor in the `package.json` that's published to npm.

Cheers.

sebald commented:
@shousper @stevegeek already noticed this in #7606, but my applied fix didn't work correctly ... my bad.

I adjusted the `copy-files` script accordingly and moved the typings back to the usual `typings` folder. Should be easier to understand for other people if the source and build have the typings in the same location.

oliviertassinari commented:
@shousper Thanks for suggesting the changes.

oliviertassinari commented on scripts/copy-files.js:
```diff
@@ -15,6 +16,16 @@ function copyFile(file) {
   }).then(() => console.log(`Copied ${file} to ${buildPath}`));
 }
 
+function copyTypings(folder) {
```
Why adding some extra logic when instead, we can put the definitions in the `/src` folder?
I'm worried about about having a more complex than needed publish script.

sebald replied:
Just putting the `.d.ts` inside the `src` folder didn't copy it. I could also do

```dif
- const files = ['README.md', 'CHANGELOG.md', 'LICENSE'];
+ const files = ['README.md', 'CHANGELOG.md', 'LICENSE', 'typings/index.d.ts'];
```

That should work too. Would you prefer that?

oliviertassinari replied:
That sounds better :). I want to reduce the overhead for new contributors. It's not obvious what this file is about. The less code branch we have, the better.

sebald replied:
Totally get that. I'm experiencing it first hand :)

oliviertassinari commented:
@sebald Should we merge the PR now or do you want to wait for the next beta this week-end?

sebald commented:
@oliviertassinari I would wait for the next beta :)

stevegeek commented:
@sebald Is it just me or is Collapse, Slide and Fade transition missing from defs?

```
declare module 'material-ui/transitions/Collapse' {
  import {TransitionCallback} from 'material-ui/internal/Transition';
  interface CollapseProps {
    in?: boolean;
    transitionDuration?: number | 'auto';
    onEnter?: TransitionCallback;
    onEntered?: TransitionCallback;
    onEntering?: TransitionCallback;
    onExit?: TransitionCallback;
    onExited?: TransitionCallback;
    onExiting?: TransitionCallback;
  }
  export default class Slide extends MaterialUI.Component<CollapseProps> {}
}


declare module 'material-ui/transitions/Slide' {
  import {TransitionCallback} from 'material-ui/internal/Transition';
  interface SlideProps {
    direction?: 'left' | 'right' | 'up' | 'down';
    enterTransitionDuration?: number;
    in?: boolean;
    leaveTransitionDuration?: number;
    offset?: string;
    onEnter?: TransitionCallback;
    onEntered?: TransitionCallback;
    onEntering?: TransitionCallback;
    onExit?: TransitionCallback;
    onExited?: TransitionCallback;
    onExiting?: TransitionCallback;
  }
  export default class Slide extends MaterialUI.Component<SlideProps> {}
}

declare module 'material-ui/transitions/Fade' {
  import {TransitionCallback} from 'material-ui/internal/Transition';
  interface FadeProps {
    enterTransitionDuration?: number;
    in?: boolean;
    leaveTransitionDuration?: number;
    offset?: string;
    onEnter?: TransitionCallback;
    onEntered?: TransitionCallback;
    onEntering?: TransitionCallback;
    onExit?: TransitionCallback;
    onExited?: TransitionCallback;
    onExiting?: TransitionCallback;
  }
  export default class Fade extends MaterialUI.Component<FadeProps> {}
}
```

sebald commented:
@stevegeek Which version of the typings are you using? I added them in 99c56580f7ed0fc3b405303c62caaa9ae4600e51

stevegeek commented:
@sebald Ah ok I must be using one from before that commit! thanks!

sebald commented:
@stevegeek You're very welcome :) Let me know if you run into problems with them!

stevegeek commented:
@sebald I think `direction` and `offset` are missing from `Slide`  https://github.com/callemall/material-ui/blob/v1-beta/src/transitions/Slide.js#L35

```
  direction?: 'left' | 'right' | 'up' | 'down';
  offset?: string;
```

Also `enterTransitionDuration` and `leaveTransitionDuration` on Slide and Fade are just `number` , not `number | 'auto'` like `transitionDuration` on Collapse (see https://github.com/callemall/material-ui/blob/v1-beta/src/transitions/Fade.js#L26 vs https://github.com/callemall/material-ui/blob/v1-beta/src/transitions/Collapse.js#L75)

stevegeek commented:
Another question, for `material-ui-icons` , what do you tihnk about adding typings to that package too? 

eg

```
// Either declare here or import SvgIconProps from main typings as all icons are implemented with SvgIcon
declare interface IconProps extends React.HTMLAttributes<{}> {
  titleAccess?: string;
  viewBox?: string;
}

declare module 'material-ui-icons/Menu' {
  export default class Menu extends React.Component<IconProps> {}
}

declare module 'material-ui-icons/ExitToApp' {
  export default class Icon extends React.Component<IconProps> {}
}

... etc for all icons. Here could write a script to generate the declarations from the files names in https://github.com/callemall/material-ui/tree/v1-beta/packages/material-ui-icons/src
```

sebald commented:
@stevegeek Thanks. Updated!

Yes, I want to add typigns for the icons, too. I just didn't have time yet. Which is kinda dump to say, because we already have typings internally ;D

We're using the following script to generate the typings:

```js
/**
 * Generate type definitions for `material-ui-icons`.
 * The generated contens will be written to
 * `src/typings/material-ui-icons.d.ts`!
 */
const { parse, resolve } = require('path');
const { EOL } = require('os');
const chalk = require('chalk');
const { writeFile } = require('fs-extra');
const globby = require('globby');
const { map, pipe } = require('ramda');
const { error } = require('./utils');
const log = console.log;

const MODULE_DIR = resolve(__dirname, '../node_modules/material-ui-icons');
const DTS_FILE = resolve(__dirname, '../src/typings/material-ui-icons.d.ts');

// Helpers
// ---------------
const normalizeFileName = name => parse(name).name;

const createTypeDefinition = name =>
  `declare module 'material-ui-icons/${name}' {
  import SvgIcon from 'material-ui/SvgIcon';
  export default class ${name} extends SvgIcon {}
}`;

const convertFilesToTypings = map(
  pipe(normalizeFileName, createTypeDefinition)
);

// Script
// ---------------
log(`\u{1f52c}  Searching for modules inside "${chalk.dim(MODULE_DIR)}".`);
globby(['*.js', '!index*'], { cwd: MODULE_DIR })
  .then(convertFilesToTypings)
  .then(typings => typings.join(`${EOL}${EOL}`))
  .then(contents => writeFile(DTS_FILE, contents, { encoding: 'utf8' }))
  .then(() => log(`\u{1F5C4}  Written typings to ${chalk.dim(DTS_FILE)}.`))
  .catch(err => {
    if (err) {
      error(err);
    }
    process.exit(1);
  });
```

Just change the `MODULE_DIR` and `DTS_FILE` and you should be good to go.

sebald commented:
Your output should look something like this: https://gist.github.com/sebald/c150a718aa05b2206dbdf1ef9e07e9d2

stevegeek commented:
@sebald Thats just perfect, thanks!!

oliviertassinari commented:
Oops, I have forgotten to merge the PR before the latest release. I will give you write access so we can work with more efficiency.

sebald commented:
@oliviertassinari  ¯\_(ツ)_/¯ happens!

Wow, thanks :-O now I am scared.

oliviertassinari commented:
Don't worry, it's gonna be all right. Whenever you want to improve the typescript coverage, go ahead. I don't know anyone better than you to take care of it.

marcusjwhelan commented:
Will there be types for Grid? @sebald

sebald commented:
@marcusjwhelan Not sure if I can follow. Do you have problems using them? Because there already are typings for `Grid`:

https://github.com/callemall/material-ui/blob/babfdb4e7ec6b158e7b1d16a5576f4efdfcf14ac/typings/index.d.ts#L587-L630

Please let me know if you run into issues with using the typings.

marcusjwhelan commented:
@sebald @oliviertassinari The Typings are not shipped with npm. I have tried `next`. There is the @types/material-ui but it is only up to date with 0.17. I am working on a large Typescript project and use this repo because I don't want to rely on a 3rd party Grid system to material-ui.

oliviertassinari commented:
We need to release these typings! I can do a quick release before the end of the week.

stevegeek commented:
@marcusjwhelan I have been using these typings for a while now and they are great so once released you can rely on them

@sebald @oliviertassinari Before releasing can we make sure to add the changes for the new `withStyles` and the removed `createStyleSheet` method

sebald commented:
@stevegeek Can you please open up an issue? (maybe even describe the change)

vyrotek commented:
@sebald I believe @stevegeek is referring to #7740 and #7730 which were just released in [beta 5](https://github.com/callemall/material-ui/releases/tag/v1.0.0-beta.5)

I'm definitely looking forward to the typings being released!

stevegeek commented:
@sebald @vyrotek Sorry just saw this message, yeah was referring to those latest changes in the API, you would like me to open an issue?

sebald commented:
@stevegeek See #7793 :)
